ts-mocha ignores code errors in tests

I have a test file which has a typo in the `import` statement, but the test still passes. Is there a way to make the test fail due to compilation errors?

Example:

```
$ mkdir ts-mocha-syntax-errors
$ cd ts-mocha-syntax-errors
$ npm init -y
$ npm install typescript expect mocha ts-mocha
$ vi tsconfig.json
$ cat tsconfig.json
{
    "compilerOptions": {
        "esModuleInterop": true
    }
}
$ vi test.ts
$ cat test.ts
import expect from 'expect';
import {Blah, Blah, Blah} from '../something';

describe('numbers', () => {
    it('fourty two is fourty two', async () => {
        expect(42).toEqual(42);
    });
});
$ node_modules/ts-mocha/bin/ts-mocha test.ts


  numbers
    ✓ fourty two is fourty two


  1 passing (8ms)
```

That `import {Blah, Blah, Blah} from '../something';` is an error. `tsc test.ts` correctly reports errors like `test.ts:2:32 - error TS2307: Cannot find module '../something'.`.

--- a/README.md
+++ b/README.md
@@ -41,7 +41,7 @@ CLI options consist of all the options of regular Mocha plus extra options below
 
 `-p, --project <value>` - relative or absolute path to a `tsconfig.json` file (equivalent of `tsc -p <value>`) [default: "./tsconfig.json"]
 
-Example:
+**Example:**
 ```bash
 ts-mocha -p src/tsconfig.json src/**/*.spec.ts
 ```
@@ -56,6 +56,15 @@ Check our test suite for a reference implementation: [Link](./test/paths/tsconfi
 **Example:**
 ```bash
 ts-mocha --paths -p src/ src/**/*.spec.ts
+```
+
+`--type-check` - feature toggle flag to enable type checking in ts-node [default: false]
+
+By default ts-mocha uses the `--transpile-only` option of ts-node to make tests run faster. Use the `--type-check` option to enable type checking in ts-node.
+
+**Example:**
+```bash
+ts-mocha --type-check -p src/ src/**/*.spec.ts
 ```
 
 ### Watch Mode:

--- a/bin/ts-mocha
+++ b/bin/ts-mocha
@@ -23,6 +23,10 @@ process.argv.slice(2).forEach(function (arg, idx, arr) {
     case '--paths':
       process.env.TS_CONFIG_PATHS = true;
       break;
+    
+    case '--type-check':
+      process.env.TS_TYPE_CHECK = true;
+      break;
 
     default:
       args.push(arg);

--- a/package.json
+++ b/package.json
@@ -16,10 +16,11 @@
     "clean": "rm -rf node_modules/",
     "reinstall": "npm run clean && npm install",
     "pretest:baseline": "npm install --peer",
-    "test:baseline": "./bin/ts-mocha -p test/baseline/tsconfig.json test/baseline/**/*.spec.ts && node test/baseline/programmatic-use-test.js",
+    "test:baseline": "./bin/ts-mocha -p test/baseline/tsconfig.json test/baseline/**/*.spec.ts test/typecheck/**/*.spec.ts && node test/baseline/programmatic-use-test.js",
     "pretest:paths": "npm install --optional",
     "test:paths": "./bin/ts-mocha --paths -p test/paths/tsconfig.json test/paths/**/*.spec.ts && node test/paths/programmatic-use-test.js",
-    "test": "npm run test:baseline & npm run test:paths",
+    "test:typecheck": "if ./bin/ts-mocha --type-check -p test/typecheck/tsconfig.json test/typecheck/**/*.spec.ts; then exit 1; fi && node test/typecheck/programmatic-use-test.js",
+    "test": "npm run test:baseline & npm run test:paths & npm run test:typecheck",
     "prepublishOnly": "npm run clean && npm install --production && npm install -D mocha && npm test"
   },
   "dependencies": {

--- a/src/index.js
+++ b/src/index.js
@@ -4,9 +4,10 @@ try {
     process.env.TS_NODE_PROJECT ||
     process.env._TS_PROJECT_PATH__ ||
     './tsconfig.json';
+  const transpileOnly = !process.env.TS_TYPE_CHECK;
   require('ts-node').register({
     project,
-    transpileOnly: true,
+    transpileOnly,
   });
   // opt-in tsconfig-paths config
   if (process.env.TS_CONFIG_PATHS) {

--- a/test/baseline/programmatic-use-test.js
+++ b/test/baseline/programmatic-use-test.js
@@ -5,6 +5,8 @@ const path = require('path');
 
 const mocha = new Mocha();
 mocha.addFile(path.resolve(__dirname, `app.spec.ts`));
+// Add a test with a compile error to prove that type checks are off by default.
+mocha.addFile(path.resolve(__dirname, `../typecheck/app.spec.ts`));
 mocha.run((failures) => {
   process.on('exit', () => {
     process.exit(failures); // exit with non-zero status if there were failures

--- a/test/typecheck/app.spec.ts
+++ b/test/typecheck/app.spec.ts
@@ -1,0 +1,13 @@
+import * as expect from 'expect';
+import app from './app';
+
+describe('Running TypeScript tests in ts-node runtime with type checks', () => {
+  describe('typecheck app module', () => {
+    it('should return the same value that was passed', () => {
+      expect(app(3)).toBe(3);
+    });
+  });
+});
+
+// Introduce a compile error.
+"not a number" as number;

--- a/test/typecheck/app.ts
+++ b/test/typecheck/app.ts
@@ -1,0 +1,3 @@
+export default function(x: number): number {
+  return x;
+}

--- a/test/typecheck/programmatic-use-test.js
+++ b/test/typecheck/programmatic-use-test.js
@@ -1,0 +1,23 @@
+process.env.TS_NODE_PROJECT = './test/paths/tsconfig.json';
+process.env.TS_TYPE_CHECK = true;
+require('../..');
+const Mocha = require('mocha');
+const path = require('path');
+const { TSError } = require('ts-node');
+
+const mocha = new Mocha();
+mocha.addFile(path.resolve(__dirname, `app.spec.ts`));
+
+try {
+  mocha.run(() => {
+    process.on('exit', () => {
+      process.exit(1); // exit with non-zero status if the tests were run at all
+    });
+  });
+} catch (error) {
+  if (error instanceof TSError) {
+    // Success, we found the compile error
+  } else {
+    throw error;
+  }
+}

--- a/test/typecheck/tsconfig.json
+++ b/test/typecheck/tsconfig.json
@@ -1,0 +1,6 @@
+{
+  "extends": "../tsconfig.base.json",
+  "compilerOptions": {
+    "baseUrl": "."
+  }
+}

